### ISSUE TITLE
The Roadmap

### ISSUE THREAD

**Issue by DBJDBJ** [Microsoft/cppwinrt#232]
This project desperately needs to detach itself from UWP, XAML, C#, .NET  and other similar issues imposed by  so called "external forces". 

I am looking forward to cppWinRT Server Side. Fully documented.

cppWinRT is pure C++ at its best.

**Comment by Scottj1s** [comment]
Can you elaborate on the server side scenarios you'd like to see supported, beyond the current ability to consume arbitrary WinRT objects on arbitrary threads?

**Comment by tim-weis** [comment]
I don't understand what's to gain from detaching a language projection of the UWP API surface from UWP. I'm also not sure, where you see any attachments to C# or .NET. C++/WinRT *is* C++. Exclusively.

**Comment by VSChawathe** [comment]
The .NET attachment has been [confirmed](https://www.reddit.com/r/cpp/comments/6vffok/cwinrt_shows_some_signs_of_life_documentation/dm16db8/). Ideally, on clean Windows installation, authoring packaged **native** UWP app ready for uploading to Windows Store Publisher account should be possible without multi platform or _managed_ runtime to prevent suffering from bloated installation.  
Declarative retained mode UI layouts with XAML are terrific from my experience wrestling with immediate mode widgets, but that's app specific. While writing XAML with C++ might amuse designer folks as fancy serialization, the design (pre)view is much more essential to engineer like me because becoming used to parsing designs mentally is hardly an option after demanding development sessions. As XAML controls are being driven by observables like Boost Synchronization Channels, surely cppwinrt might [rectify](http://kirkshoop.github.io/2016/12/16/rxcpp_observable_as_a_coroutine_range.html) them, instead of INotifyPropertyChanged based MV* IoC coding pattern's over wiring and code bloat.. Rather than removing XAML from this UWP projection, let's hope it keeps improvising as cppwinrt/XAML. It's as intuitive as CSV data stores, YAML (de)serialization, GraphQL, ..  
**Pure** C++ at its best will require accessing framebuffer by addressing raw memory effectively rewriting the implementation divided between OS (development platform) and hardware vendor, as the 2D graphics proposal is awaiting being standardized as part of C++ by ISO.

**Comment by BrentRe** [comment]
The UWP API has nothing to do with .NET. You claim that it is is incorrect. Even at the link you mention, Andrew simply states that the C++/CX compiler uses one .NET DLL to write metadata. Just because the CX compiler uses the DLL doesn't mean that your compiled application use it - and it does not.

Additionally, **authoring** a packaged native UWP application creates a runtime with what dependencies required by your programming language plus application code. For example, a C++ UWP application typically will have a dependency on the C Runtime. A C# UWP application is compiled to native code and will typically have a dependency on the .NET Native runtime and the C Runtime. A JavaScript UWP application has a dependency on the JavaScript engine.

Neither produces a "bloated" installation IMO. Your application gets the dependencies it requires.

**Comment by VSChawathe** [comment]
Just to clarify, download payloads the size of VS installers are much more affordable at my end on slower connections, so the weakest link of the VC++ development cycle for cppwinrt is the bloated installer. As I've mentioned "authoring" against the UWP API, so bloated installation is about how lean & short VS installation can get for cppwinrt UWP store ready app development only, when it arrives side by side with other projects in VS. I never intended to suggest anything about the "produce".

**Comment by DBJDBJ** [comment]
Gentlemen, if I may :) this thread is called "The Roadmap". What I propose is to kindly ask the team to try and arrive to the cppWinRt which will have nothing to do with any of the technologies/mechanism I have enumerated (not comprehensively).  

Bits and pieces I assume already in. They just need to be decoupled, documented and used. Perhaps something like cppWinRT core/kernel ... I also assume this is in the pipeline, but I do not know. As there are no any architectural artifacts (diagrams etc.) I know about.

In case somebody might get confused here: "server side" code needs none of them (xaml,upw, etc..) and this is why I am introducing this term to clarify what I mean. The code that runs on the server. No human interaction, no ui, no gui.  The contemporary examples perhaps easier to understand are: web services implementations.  That kind of software. COM servers and a such .. Using WIN32, WMI, AD and a such. Also implementing Services ... Web servers. Databases. 

Implementation is (it seems) brilliant. That is not the issue. I (we?) need the Roadmap to know where is this all going and when.

**Comment by tim-weis** [comment]
I don't understand, what you are proposing, though. If you don't need a UI/GUI, don't implement a UI/GUI. If you don't want to use XAML, don't use XAML (which is easy, due to the limited XAML support of the current C++/WinRT library). If you don't want to target the UWP, don't target the UWP.

If you want to access the Windows Runtime from a desktop application or Windows service, simply access those Windows Runtime types you need (the majority of the Windows Runtime types are accessible from desktop applications, with the notable exception of the GUI classes, which you have no interest in using anyway).

As it stands, C++/WinRT currently offers everything you are asking for (minus documentation). Most of the features that get added are of little or no interest to you given your requirements. You are all set to go, right now.

Unless I'm totally missing the issue you were trying to report.

**Comment by DBJDBJ** [comment]
@tim-weis  The Roadmap will at least tell us what and when will be implemented. And most importantly why.  It should have been done first. But it is not published. 

Pretty succinct text is [here](https://en.wikipedia.org/wiki/Technology_roadmap).

There are no architectural artifacts. Just sea of c++17 to marvel at...

**Comment by andrew7webb** [comment]
_This is what the readme.md says:_

                             Timeline
Following is an overview roadmap for upcoming releases:
The C++/WinRT headers and the cppwinrt.exe compiler itself ... <waiting for the cheers to die down> ... will ship in the Windows SDK. Expect to first see them in a Insider Preview build of the Windows SDK in late summer/early fall.
We are updating the XAML compiler to produce C++/WinRT code behind files.
We are adding C++/WinRT project support to Visual Studio 2017.
C++/WinRT will stay current with C++ standards and require a modern and up-to-date compiler, generally the latest release of Visual Studio. Users requiring support for older compilers will need to sync to an earlier commit or tag.


_I'm happy with that timeline.   #1 on my wish list after those items are completed is a C++/WinRT Programmers guide targeted at C++ programmers without C++/CX or Windows Runtime API skills._

**Comment by tim-weis** [comment]
I can totally sympathize with the idea of having an official roadmap. For *my* use cases. I'm still struggling to understand, why *you* are asking for one. You don't have to wait for any milestones. Unless there is something you haven't told us yet, the current release of the C++/WinRT library provides everything *you* need. Specifically, as @Scottj1s pointed out, *"the [...] ability to consume arbitrary WinRT objects on arbitrary threads"*. The only Windows Runtime types you cannot instantiate on arbitrary threads are those you have no interest in using anyway.

I would agree, that documentation is the weakest aspect of C++/WinRT as of now. And still, you haven't made a convincing point for a roadmap, given your intended uses case (server-side code). Whatever is happening in a future update has no impact on your current ability, to fully use C++/WinRT (breaking changes aside).

I can't help but think, that I missed a significant detail. Correct me if this is the case.

**Comment by VSChawathe** [comment]
I believe we can contribute to docs.microsoft.com if something is really needed via pull request on GitHub. For queries where less expertise than the cppwinrt team members themselves will suffice, I'd enjoy preparing the content for them. VS2017 installer shows 3 tabs as Workloads, Individual components, and Language packs. I am just waiting with my breath on hold for cppwinrt to be amongst the workloads, with fewer individual components (why install C#, VB at once, if I will visit them only once awhile). 
 As for roadmap, as C++ itself is moving target, so speculating roadmap maybe overburdening.

**Comment by DBJDBJ** [comment]
Ok I am not into debating "why the Roadmap" or how does one look etc.  We simply have not seen one yet.  Of course there are commercial boundaries dictating should one be made public or not. 

@tim-weis I am not even going to attempt to (what it amounts to) reverse engineer cppWinRT just to be able to use it on the server side.  I will simply not dedicate any resource to cppWinRT untill we have "cppWinRT Programmers guide" with only pure C++17 in it. Until then we can just speculate.

@VSChawathe  cppWinRT (and the team) has no obvious problems with a "moving target" neither do I.

As for XAML, UWP and a such ... if it happens I need some kind of GUI, I follow the logic of the very smooth ["Code" editor](https://code.visualstudio.com/) and the MSFT team behind it.

Many thanks all for your thoughts :)
